core/src/main/java/io/undertow/server/DetachableStreamSinkConduit.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2024 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.undertow.server;
+
+import io.undertow.UndertowMessages;
+import org.xnio.XnioIoThread;
+import org.xnio.XnioWorker;
+import org.xnio.channels.StreamSourceChannel;
+import org.xnio.conduits.StreamSinkConduit;
+import org.xnio.conduits.WriteReadyHandler;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * When this conduit is considered detached it will no longer forward calls to the delegate.
+ */
+abstract class DetachableStreamSinkConduit implements StreamSinkConduit {
+
+    private final StreamSinkConduit delegate;
+
+    DetachableStreamSinkConduit(StreamSinkConduit delegate) {
+        this.delegate = delegate;
+    }
+
+    protected abstract boolean isFinished();
+
+    @Override
+    public long transferFrom(FileChannel src, long position, long count) throws IOException {
+        if (isFinished()) {
```

Comment:
Have you done any performance testing on repeated calls to `isFinished()`? Is there any way to memoize/cache the check? Am I optimizing prematurely? :)

Reply:
This didn't incur a measurable cost in my benchmarks. This `isFinished` check is identical to the `isFinished` check within `HttpServerExchange.WriteDispatchChannel` which handles a similar check at a slightly different layer.
This shouldn't have much performance impact because the underlying check is `state & 2048 == 2048` where `state` is a standard non-volatile field.
I considered splitting the existing `WriteDispatchChannel` implementation so that we only need to do this check in the new conduit implementation, however that seemed riskier, and this way we do still short-circuit wrapper logic after the response has completed.